File: test/CSharp/CodeCracker.Test/Usage/JsonNetAnalyzerTests.cs
using System.Threading.Tasks;
using CodeCracker.CSharp.Usage;
using Microsoft.CodeAnalysis;
using Microsoft.CodeAnalysis.Diagnostics;
using Microsoft.CodeAnalysis.Testing;
using Xunit;

namespace CodeCracker.Test.CSharp.Usage
{
    public class JsonNetAnalyzerTests : CodeFixVerifier
    {
        private const string TestCode = @"
using System;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;
namespace ConsoleApplication1
{{
    class Person
    {{
        public Person()
        {{
            {0}
        }}
    }}
}}";

        [Fact]
        public async Task IfDeserializeObjectIdentifierFoundAndJsonTextIsIncorrectCreatesDiagnostic()
        {
            var test = string.Format(TestCode, @"Newtonsoft.Json.JsonConvert.DeserializeObject<Person>(""foo"");");
            await VerifyCSharpDiagnosticAsync(test, CreateDiagnosticResult(11, 67));
        }

        [Fact]
        public async Task IfAbbreviatedDeserializeObjectIdentifierFoundAndJsonTextIsIncorrectCreatesDiagnostic()
        {
            var test = string.Format(TestCode, @"JsonConvert.DeserializeObject<Person>(""foo"");");
            await VerifyCSharpDiagnosticAsync(test, CreateDiagnosticResult(11,51));
        }

        [Fact]
        public async Task IfDeserializeObjectIdentifierFoundAndJsonTextIsCorrectDoesNotCreatesDiagnostic()
        {
            var test = string.Format(TestCode, @"Newtonsoft.Json.JsonConvert.DeserializeObject<Person>(""{""name"":""foo""}"");");
            await VerifyCSharpHasNoDiagnosticsAsync(test);
        }

        [Fact]
        public async Task IfAbbreviateDeserializeObjectIdentifierFoundAndJsonTextIsCorrectDoesNotCreatesDiagnostic()
        {
            var test = string.Format(TestCode, @"JsonConvert.DeserializeObject<Person>(""{""name"":""foo""}"");");
            await VerifyCSharpHasNoDiagnosticsAsync(test);
        }

        [Fact]
        public async Task IfJObjectParseIdentifierFoundAndJsonTextIsIncorrectCreatesDiagnostic()
        {
            var test = string.Format(TestCode, @"Newtonsoft.Json.Linq.JObject.Parse(""foo"");");
            await VerifyCSharpDiagnosticAsync(test, CreateDiagnosticResult(11, 48));
        }

        [Fact]
        public async Task IfAbbreviatedJObjectParseIdentifierFoundAndJsonTextIsIncorrectCreatesDiagnostic()
        {
            var test = string.Format(TestCode, @"JObject.Parse(""foo"");");
            await VerifyCSharpDiagnosticAsync(test, CreateDiagnosticResult(11, 27));
        }

        [Fact]
        public async Task IfJObjectParseIdentifierFoundAndJsonTextIsCorrectDoesNotCreatesDiagnostic()
        {
            var test = string.Format(TestCode, @"JObject.Parse(""{""name"":""foo""}"");");
            await VerifyCSharpHasNoDiagnosticsAsync(test);
        }
        //
        [Fact]
        public async Task IfJArrayParseIdentifierFoundAndJsonTextIsIncorrectCreatesDiagnostic()
        {
            var test = string.Format(TestCode, @"Newtonsoft.Json.Linq.JArray.Parse(""foo"");");
            await VerifyCSharpDiagnosticAsync(test, CreateDiagnosticResult(11, 47));
        }

        [Fact]
        public async Task IfAbbreviatedJArrayParseIdentifierFoundAndJsonTextIsIncorrectCreatesDiagnostic()
        {
            var test = string.Format(TestCode, @"JArray.Parse(""foo"");");
            await VerifyCSharpDiagnosticAsync(test, CreateDiagnosticResult(11, 26));
        }

        [Fact]
        public async Task IfJArrayParseIdentifierFoundAndJsonTextIsCorrectDoesNotCreatesDiagnostic()
        {
            var test = string.Format(TestCode, @"JArray.Parse(""{""name"":""foo""}"");");
            await VerifyCSharpHasNoDiagnosticsAsync(test);
        }

        private static DiagnosticResult CreateDiagnosticResult(int line, int column) {
            return new DiagnosticResult(DiagnosticId.JsonNet.ToDiagnosticId(), DiagnosticSeverity.Error)
                .WithLocation(line, column)
                .WithMessage("Unexpected end when reading JSON. Path '', line 1, position 3.");
        }

        protected override DiagnosticAnalyzer GetDiagnosticAnalyzer() => new JsonNetAnalyzer();
    }
}

File: test/VisualBasic/CodeCracker.Test/Usage/JsonNetAnalyzerTests.vb
Imports CodeCracker.VisualBasic.Usage
Imports Microsoft.CodeAnalysis.Diagnostics
Imports Microsoft.CodeAnalysis.Testing
Imports Xunit

Namespace Usage
    Public Class JsonNetAnalyzerTests
        Inherits CodeFixVerifier

        Private Const TestCode = "
Imports System
Imports Newtonsoft.Json
Imports Newtonsoft.Json.Linq
Namespace ConsoleApplication1
    Class Person
        Public Sub New()
            {0}
        End Sub
    End Class
End Namespace"

        Private Shared Function CreateDiagnosticResult(line As Integer, column As Integer) As DiagnosticResult
            Return New DiagnosticResult(DiagnosticId.JsonNet.ToDiagnosticId(), Microsoft.CodeAnalysis.DiagnosticSeverity.Error) _
                .WithLocation(line, column) _
                .WithMessage("Unexpected end when reading JSON. Path '', line 1, position 3.")
        End Function

        Protected Overrides Function GetDiagnosticAnalyzer() As DiagnosticAnalyzer
            Return New JsonNetAnalyzer
        End Function

        <Fact>
        Public Async Function IfDeserializeObjectIdentifierFoundAndJsonTextIsIncorrectCreatesDiagnostic() As Task
            Dim test = String.Format(TestCode, "Newtonsoft.Json.JsonConvert.DeserializeObject(of Person)(""foo"")")
            Await VerifyBasicDiagnosticAsync(test, CreateDiagnosticResult(8, 70))
        End Function

        <Fact>
        Public Async Function IfAbbreviatedDeserializeObjectIdentifierFoundAndJsonTextIsIncorrectCreatesDiagnostic() As Task
            Dim test = String.Format(TestCode, "JsonConvert.DeserializeObject(of Person)(""foo"")")
            Await VerifyBasicDiagnosticAsync(test, CreateDiagnosticResult(8, 54))
        End Function

        <Fact>
        Public Async Function IfDeserializeObjetIdentifierFoundAndJsonTextIsCorrectDoesNotCreateDiagnostic() As Task
            Dim test = String.Format(TestCode, "Newtonsoft.Json.JsonConvert.DeserializeObject(of Person)(""{""""name"""":""""foo""""}"")")
            Await VerifyBasicHasNoDiagnosticsAsync(test)
        End Function

        <Fact>
        Public Async Function IfAbbreviatedDeserializeObjetIdentifierFoundAndJsonTextIsCorrectDoesNotCreateDiagnostic() As Task
            Dim test = String.Format(TestCode, "JsonConvert.DeserializeObject(of Person)(""{""""name"""":""""foo""""}"")")
            Await VerifyBasicHasNoDiagnosticsAsync(test)
        End Function

        <Fact>
        Public Async Function IfJObjectParseIdentifierFoundAndJsonTextIsIncorrectCreatesDiagnostic() As Task
            Dim test = String.Format(TestCode, "Newtonsoft.Json.Linq.JObject.Parse(""foo"")")
            Await VerifyBasicDiagnosticAsync(test, CreateDiagnosticResult(8, 48))
        End Function

        <Fact>
        Public Async Function IfAbbreviatedJObjectParseIdentifierFoundAndJsonTextIsIncorrectCreatesDiagnostic() As Task
            Dim test = String.Format(TestCode, "JObject.Parse(""foo"")")
            Await VerifyBasicDiagnosticAsync(test, CreateDiagnosticResult(8, 27))
        End Function

        <Fact>
        Public Async Function IfJObjectParseFoundAndJsonTextIsCorrectDoesNotCreateDiagnostic() As Task
            Dim test = String.Format(TestCode, "JObject.Parse(""{""""name"""":""""foo""""}"")")
            Await VerifyBasicHasNoDiagnosticsAsync(test)
        End Function

        <Fact>
        Public Async Function IfJArrayParseIdentifierFoundAndJsonTextIsIncorrectCreatesDiagnostic() As Task
            Dim test = String.Format(TestCode, "Newtonsoft.Json.Linq.Jarray.Parse(""foo"")")
            Await VerifyBasicHasNoDiagnosticsAsync(test)
        End Function

        <Fact>
        Public Async Function IfAbbreviatedJArrayParseIdentifierFoundAndJsonTextIsIncorrectCreatesDiagnostic() As Task
            Dim test = String.Format(TestCode, "Jarray.Parse(""foo"")")
            Await VerifyBasicHasNoDiagnosticsAsync(test)
        End Function

        <Fact>
        Public Async Function IfJArrayParseFoundAndJsonTextIsCorrectDoesNotCreateDiagnostic() As Task
            Dim test = String.Format(TestCode, "JArray.Parse(""{""""name"""":""""foo""""}"")")
            Await VerifyBasicHasNoDiagnosticsAsync(test)
        End Function

    End Class
End Namespace